File: src/main/java/no/fint/explorer/service/MetricsService.java
package no.fint.explorer.service;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tag;
import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.health.Health;
import no.fint.explorer.model.Asset;
import no.fint.explorer.model.SseOrg;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

@Slf4j
@Service
public class MetricsService {

    private final static String HEALTHY = "APPLICATION_HEALTHY";
    private final static String UNHEALTHY = "APPLICATION_UNHEALTHY";
    private final static String HEALTH_METRIC = "fint.core.health";
    private final static String CACHE_METRIC = "fint.core.cache";
    private final static String ADAPTER_COUNT_METRIC = "fint.core.adapter.count";
    private final static String ADAPTER_EVENTS_METRIC_TOTAL = "fint.core.adapter.events.total";

    private final MeterRegistry meterRegistry;
    private final AssetService assetService;

    public MetricsService(MeterRegistry meterRegistry, AssetService assetService) {
        this.meterRegistry = meterRegistry;
        this.assetService = assetService;
    }

    @Scheduled(initialDelayString = "${kubernetes.initial-delay}", fixedDelayString = "${kubernetes.fixed-delay}")
    public void update() {

        assetService.update();

        log.info("Start updating metrics...");
        assetService.getAssets().toStream()
                .forEach(asset -> {
                    updateHealthMetric(asset);
                    updateCacheMetric(asset);
                    updateAdapterMetric(asset);
                });
        log.info("Finished updating metrics");
    }

    private void updateHealthMetric(Asset asset) {
        asset.getComponents()
                .forEach(component -> {
                    meterRegistry.gauge(HEALTH_METRIC,
                            Arrays.asList(Tag.of("asset", asset.getId()), Tag.of("component", component.getId())),
                            new AtomicInteger(getHealthStatus(component.getHealth())));
                });
    }

    private void updateCacheMetric(Asset asset) {
        asset.getComponents()
                .forEach(component -> {
                    component
                            .getCache()
                            .forEach(cacheEntry -> {
                                meterRegistry.gauge(CACHE_METRIC,
                                        Arrays.asList(Tag.of("asset", asset.getId()), Tag.of("component", component.getId()), Tag.of("entity", cacheEntry.getName())),
                                        new AtomicInteger(cacheEntry.getSize()));
                            });
                });
    }

    private void updateAdapterMetric(Asset asset) {
        asset.getComponents()
                .forEach(component -> {
                    meterRegistry.gauge(ADAPTER_EVENTS_METRIC_TOTAL,
                            Arrays.asList(Tag.of("asset", asset.getId()), Tag.of("component", component.getId())),
                            new AtomicInteger(component.getClients().stream().mapToInt(SseOrg.SseClient::getEvents).sum()));

                    meterRegistry.gauge(ADAPTER_COUNT_METRIC,
                            Arrays.asList(Tag.of("asset", asset.getId()), Tag.of("component", component.getId())),
                            new AtomicInteger(component.getClients().size()));
                });
    }

    private int getHealthStatus(List<Health> healthList) {
        return healthList
                .stream()
                .map(Health::getStatus)
                .filter(HEALTHY::equals)
                .findFirst()
                .orElse(UNHEALTHY)
                .equals(HEALTHY) ? 1 : 0;
    }
}

File: src/main/java/no/novari/fint/explorer/service/MetricsService.kt
package no.novari.fint.explorer.service

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Tag
import no.fint.event.model.health.Health
import no.novari.fint.explorer.model.Asset
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

@Service
class MetricsService(
    private val meterRegistry: MeterRegistry,
    private val assetService: AssetService,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    private val gauges = ConcurrentHashMap<String, AtomicInteger>()
    private val longGauges = ConcurrentHashMap<String, AtomicLong>()

    @Scheduled(initialDelayString = "\${fint.poll.initial-delay}", fixedDelayString = "\${fint.poll.fixed-delay}")
    fun update() {
        assetService.update()

        log.info("Start updating metrics...")
        assetService.getAssets().forEach { asset ->
            updateHealthMetric(asset)
            updateCacheMetric(asset)
            updateAdapterMetric(asset)
        }
        log.info("Finished updating metrics")
    }

    private fun updateHealthMetric(asset: Asset) {
        asset.components.forEach { component ->
            updateMetric(
                HEALTH_METRIC,
                listOf(Tag.of("asset", asset.id), Tag.of("component", component.id)),
                getHealthStatus(component.health),
            )
        }
    }

    private fun updateCacheMetric(asset: Asset) {
        asset.components.forEach { component ->
            component.cache.forEach { cacheEntry ->
                val tags = listOf(
                    Tag.of("asset", asset.id),
                    Tag.of("component", component.id),
                    Tag.of("entity", cacheEntry.name),
                )
                updateMetric(CACHE_METRIC, tags, cacheEntry.size)
                cacheEntry.lastUpdated?.let {
                    updateLongMetric(CACHE_LAST_UPDATED_METRIC, tags, it.toEpochSecond())
                }
            }
        }
    }

    private fun updateAdapterMetric(asset: Asset) {
        asset.components.forEach { component ->
            updateMetric(
                ADAPTER_EVENTS_METRIC_TOTAL,
                listOf(Tag.of("asset", asset.id), Tag.of("component", component.id)),
                component.clients.sumOf { it.events },
            )
            updateMetric(
                ADAPTER_CONNECTIONS_METRIC,
                listOf(Tag.of("asset", asset.id), Tag.of("component", component.id)),
                component.clients.size,
            )
        }
    }

    private fun getHealthStatus(healthList: List<Health>): Int =
        if (healthList.any { it.status == HEALTHY }) 1 else 0

    private fun updateMetric(metricType: String, tags: List<Tag>, value: Int) {
        val gaugeId = metricType + "-" + tags.joinToString("-") { it.value }
        gauges.computeIfPresent(gaugeId) { _, gauge -> gauge.apply { set(value) } }
        gauges.putIfAbsent(gaugeId, meterRegistry.gauge(metricType, tags, AtomicInteger(value)))
    }

    private fun updateLongMetric(metricType: String, tags: List<Tag>, value: Long) {
        val gaugeId = metricType + "-" + tags.joinToString("-") { it.value }
        longGauges.computeIfPresent(gaugeId) { _, gauge -> gauge.apply { set(value) } }
        longGauges.putIfAbsent(gaugeId, meterRegistry.gauge(metricType, tags, AtomicLong(value)))
    }

    companion object {
        private const val HEALTHY = "APPLICATION_HEALTHY"
        private const val HEALTH_METRIC = "fint.core.health"
        private const val CACHE_METRIC = "fint.core.cache"
        private const val CACHE_LAST_UPDATED_METRIC = "fint.core.cache.last.updated.seconds"
        private const val ADAPTER_CONNECTIONS_METRIC = "fint.core.adapter.connections"
        private const val ADAPTER_EVENTS_METRIC_TOTAL = "fint.core.adapter.events.total"
    }
}
